src/plugin/debugger/mhpdbgc.c

```diff
@@ -1051,80 +1051,92 @@ static void mhp_dump_to_file(int argc, char * argv[])
 
 static int is_valid_program_name(const char *s)
 {
+  size_t len = strlen(s);
   const char *p;
 
   for (p = s; *p; p++) {
     if (iscntrlDOS(*p))
       return 0;
   }
-  return (s[0] != 0); // at least one character long
+  return (len > 0 && len <= 8);
 }
 
-static const char *get_name_from_mcb(struct MCB *mcb, int *is_lnk)
-{
-  const char *dos = "DOS", *fre = "FREE", *lnk = "LINK";
-  static char name[9];
+static int get_mcb_name_segment_psp(uint16_t seg, uint16_t off, char *out, int osize);
 
+static int get_name_from_mcb(struct MCB *mcb, int *is_lnk, char *out, int osize)
+{
   if (is_lnk)
     *is_lnk = 0;
-  if (mcb->owner_psp == 0)
-    return fre;
+  if (mcb->owner_psp == 0) {
+    strlcpy(out, "FREE", osize);
+    return 1;
+  }
+
   if (mcb->owner_psp == 8) {
     if (strcmp(mcb->name, "SC") == 0) {
       if (is_lnk)
         *is_lnk = 1;
-      return lnk;
+      strlcpy(out, "LINK", osize);
+      return 1;
     }
-    return dos;
+    strlcpy(out, "DOS", osize);
+    return 1;
   }
-  snprintf(name, sizeof name, "%s", mcb->name);
-  if (!is_valid_program_name(name))
-    snprintf(name, sizeof name, "%05d", mcb->owner_psp);
 
-  return name;
+  if (is_valid_program_name(mcb->name)) {
```

Comment:
I propose the different strategy.
The valid name can always be found
by that formula:
((struct mcb FAR *)MK_FP(mcb->m_psp - 1, 0))->m_name

Reply:
How so, if m_name points to crap like MS-DOS does? Is there a condition where testing owner_psp somehow can show that it's the ultimate ancestor(i.e. the program), but I don't to ascend all the way back to command.com?

Reply:
I am quite sure the name is always
set correctly on an mcb that precedes
the master psp. We rely on that in
builtins IIRC.

Reply:
DOS 3x doesn't set it at all, hence this code was originally copied from X window title setting.

So imagine I'm walking the chain of MCBs, how do I tell that the current one is the ultimate ancestor and I should trust its name field, or else look to its parent (owner_psp - 1) and use its name field instead?

Also is it possible that an MCB has a grandparent, i.e. mcb->(owner_psp -1)->(owner_psp -1)->name?

Reply:
> DOS 3x doesn't set it at all

Yep but I wonder if it leaves crap,
or properly blanks? Anyway, why
to care about DOS <3x? :)

> its name field, or else look to its parent (owner_psp - 1) and use its name field instead?

You can use the above formula
unconditionally. Even if its the master
mcb, by doing psp-1 you'll get to
itself, and the name will be valid
because it was from the beginning.
Of course you can add additional
decorations, like asterisk for a master
mcbs, or alike. :) In that case you
will need the test, but otherwise not
(the test is: FP_SEG(mcb)==mcb->m_psp-1)

> Also is it possible that an MCB has a grandparent

Theoretically it may be possible by the
use of an undocumented function 0x55
but normally that never happens.
Of course you can still apply your validator
func, but it is much better to only use it to
detect an errors, rather than for a routine
usage. What if the program name uses some
non-437 codepage? You may mistreat that
as an invalid name.

Reply:
An MCB with its own memory block address (ie MCB segment + 1) has a valid name, ASCIZ or up to 8 bytes in the name field, for MS-DOS 4+. An MCB with a special value like 8 can have a valid name composed of two or eight bytes. I consider it as having up to two bytes if the first byte is a capital "S", up to eight else. (MS-DOS uses "SC" or "SD" and does not initialise the third byte in this case.) In addition, I consider all owner values < 50h as "special" because there cannot be MCBs in the IVT or BDA.

(RxDOS uses owner = 8 and name = "S" followed by a null byte for its S MCBs which are used for all sorts of system data. There are a few more fields in the six bytes of the name field following the "S" and null.)

Reply:
Thanks both, I've just pushed a revised patch that I think addresses your concerns.